packages/ember/addon/index.ts

```diff
@@ -0,0 +1,57 @@
+import { defer, reject } from 'rsvp';
+import { run } from '@ember/runloop';
+import EmberObject from '@ember/object';
+import { registerSystem } from '@milestones/core';
+
+declare const require: {
+  (module: string): any; // eslint-disable-line
+  has(module: string): boolean;
+};
+
+registerSystem({ run, defer });
+
+if (require.has('ember-concurrency')) {
+  const getRunningInstance = require('ember-concurrency/-task-instance').getRunningInstance;
+  const taskMacro = require('ember-concurrency').task;
+  class TaskHost extends EmberObject.extend({
+    started: false,
+    realPromise: null as unknown,
+    milestoneTask: taskMacro(function(this: TaskHost, promise: Promise<unknown>) {
+      return {
+        next: () => {
+          if (!this.started) {
+            this.started = true;
+            return { value: promise, done: false };
+          } else {
+            return { value: this.realPromise, done: true };
+          }
+        },
+      };
+    }),
+  }) {}
+
+  registerSystem({
+    defer() {
+      let { resolve, promise } = defer();
+      let obj = TaskHost.create();
+      let task = obj.get('milestoneTask');
+      let dfd = {
+        promise: (getRunningInstance() ? task.linked() : task).perform(promise),
+        cancel(reason: unknown) {
```

Comment:
I want to make sure I'm reading this right....

By default, the milestones system handles cancelation by calling `this._deferred.cancel`, which is not supported by native `Promise`. So in order to allow for milestone cancelation, one must use `registerSystem` and provide a `defer` function that returns an object with a `.cancel` function.

And you moved the error message about cancelation support from the [ember-specific milestone implementation](https://github.com/salsify/ember-milestones/blob/bc8fc6570baa1d7ab3209969854f9b3101be2ea6/addon/-private/defer.ts#L58-L60) to the [milestone-handler itself](https://github.com/salsify/ember-milestones/pull/15/files#diff-49646b461ff0129fa79826229c013260R34).

Reply:
That's right! This was the simplest way I could come up with to keep the core system reasonable while still allowing for ember-concurrency to be plugged into it.

In principle this would allow you to plug milestones into a totally separate system with its own idea of cancelation, like [`redux-saga`](https://github.com/redux-saga/redux-saga).